proposals/0000-numeric-underscores.rst

```diff
@@ -0,0 +1,173 @@
+This proposal
+~~~~~~~~~~~~~
+I propose an extension to the existing syntax of numeric literals.
+
+Current syntax:
+
+.. code-block:: none
+
+    decimal     →  digit{digit}
+    octal       →  octit{octit}
+    hexadecimal →  hexit{hexit}
+    binary      →  binit{binit}
+
+New syntax (this proposal):
+
+.. code-block:: none
+
+    decimal     →  digit{[_ | digit]}
+    octal       →  octit{[_ | octit]}
+    hexadecimal →  hexit{[_ | hexit]}
+    binary      →  binit{[_ | binit]}
+
+    Underscores (_) in numeric literals are simply ignored.
+
+Current specification of numeric literals
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Detail of current specification in `Haskell 2010 Language Report <https://www.haskell.org/onlinereport/haskell2010/haskellch2.html#x7-190002.5>`_, chapter 2:
+
+.. code-block:: none
+
+    decimal     →  digit{digit}
+    octal       →  octit{octit}
+    hexadecimal →  hexit{hexit}
+
+    integer → decimal
+             | 0o octal | 0O octal
+             | 0x hexadecimal | 0X hexadecimal
+
+    float → decimal . decimal [exponent]
+           | decimal exponent
+
+    exponent → (e | E) [+ | -] decimal
+
+    digit    →  ascDigit | uniDigit
+    ascDigit →  0 | 1 | … | 9
+    uniDigit →  any Unicode decimal digit
+    octit    →  0 | 1 | … | 7
+    hexit    →  digit | A | … | F | a | … | f
+
+Detail of current specification in `BinaryLiterals <https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html?highlight=binaryliterals#ghc-flag--XBinaryLiterals>`_ language extension in my understanding:
+
+.. code-block:: none
+
+    binary      →  binit{binit}
+    binit       →  0 | 1
+
+    integer → decimal
+             | 0o octal | 0O octal
+             | 0x hexadecimal | 0X hexadecimal
+             | 0b binary | 0B binary
+
+Examples
+--------
+The followings are examples of this proposal:
+
+.. code-block:: none
+
+    -- decimal
+    million    = 1_000_000
+    billion    = 1_000_000_000
+    lightspeed = 299_792_458
+    version    = 8_04_1
+    date       = 2017_12_31
+
+    -- hexadecimal
+    red_mask = 0xff_00_00
+    size1G   = 0x3fff_ffff
+
+    -- binary
+    bit8th   = 0b01_0000_0000
+    packbits = 0b1_11_01_0000_0_111
+    bigbits  = 0b1100_1011__1110_1111__0101_0011
+
+    -- float
+    pi       = 3.141_592_653_589_793
+    faraday  = 96_485.332_89
+    avogadro = 6.022_140_857e+23
+
+    -- function
+    isUnderMillion = (< 1_000_000)
+
+    clip64M x
+        | x > 0x3ff_ffff = 0x3ff_ffff
+        | otherwise = x
+
+    test8bit x = (0b01_0000_0000 .&. x) /= 0
+
+Effect and Interactions
+-----------------------
+I believe that this proposal will improve the readability, quality and expressiveness of native numeric literals without degrading performance.
+
+Costs and Drawbacks
+-------------------
+* Implementation costs are mostly related to lexers.
+* Maintenance costs are related to compatibility. Compatibility can be handled with language extension like ``NumericUnderscores``.
```

Comment:
You should probably make the proposal right away suggest this language extension, as a language extensions is required for such changes.

Reply:
We could fold it into `BinaryLiterals` or something like this, but yes, I generally agree that a new extension is probably best.

Reply:
Thanks for the feedback.
I agree with the language extension.
I will update the proposal for specifying language extension.

Reply:
I explicitly described the language extension at [this commit][1].

[1]: https://github.com/ghc-proposals/ghc-proposals/pull/76/commits/ca707d6fb8b3d57c5307283a00c00af6f4cd1d62